File: internship_backend_apr25/jury_solutions/python/a.py
allCardsCount = 52
allSuits = set("CDHS")
allRanks = set("23456789TJQKA")


def solve():
    groups_count = list(map(int, input().split()))

    probability1 = calculate_probability(groups_count[0], groups_count[1])
    probability2 = calculate_probability(groups_count[2], groups_count[3])

    if probability1 > probability2:
        print(probability1)
    else:
        print(probability2)


def calculate_probability(removed_groups_count, sought_groups_count):
    removed_cards = set()
    for i in range(removed_groups_count):
        group = set(input())
        removed_cards.update(get_cards(group))

    sought_cards = set()
    for i in range(sought_groups_count):
        group = set(input())
        sought_cards.update(get_cards(group))

    if len(removed_cards) == allCardsCount:
        return 0.0
    removed_cards_count = len(removed_cards)
    removed_cards.intersection_update(sought_cards)
    cards_intersection = len(removed_cards)
    return (len(sought_cards) - cards_intersection) / (allCardsCount - removed_cards_count)


def get_cards(group):
    current_suits = set(allSuits)
    current_suits.intersection_update(group)
    if len(current_suits) == 0:
        current_suits = allSuits

    current_ranks = group.difference(current_suits)
    if len(current_ranks) == 0:
        current_ranks = allRanks

    cards = set()
    for rank in current_ranks:
        for suit in current_suits:
            cards.add(rank + suit)

    return cards


if __name__ == '__main__':
    solve()

File: internship_backend_apr25/jury_solutions/cs/a.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApplication
{
    internal class Program
    {
        private const int AllCardsCount = 52;

        private static readonly HashSet<char> AllSuits = new HashSet<char>()
        {
            'C', 'D', 'H', 'S'
        };

        private static readonly HashSet<char> AllRanks = new HashSet<char>()
        {
            '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A'
        };
        
        public static void Main(string[] args)
        {
            var groupsCount = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            
            var probability1 = CalculateProbability(groupsCount[0], groupsCount[1]);
            var probability2 = CalculateProbability(groupsCount[2], groupsCount[3]);
            
            if (probability1 > probability2)
                Console.WriteLine(probability1);
            else
                Console.WriteLine(probability2);
        }

        private static double CalculateProbability(int removedGroupsCount, int soughtGroupsCount)
        {
            var removedCards = new HashSet<string>();
            for (int i = 0; i < removedGroupsCount; i++)
            {
                var group = new HashSet<char>(Console.ReadLine());
                removedCards.UnionWith(GetCards(group));
            }
            
            var soughtCards = new HashSet<string>();
            for (int i = 0; i < soughtGroupsCount; i++)
            {
                var group = new HashSet<char>(Console.ReadLine());
                soughtCards.UnionWith(GetCards(group));
            }
            
            if (removedCards.Count == AllCardsCount)
                return 0.0;
            var removedCardsCount = removedCards.Count;
            
            removedCards.IntersectWith(soughtCards);
            var cardsIntersection = removedCards.Count;
        
            return (double)(soughtCards.Count - cardsIntersection) / (AllCardsCount - removedCardsCount);
        }
        
        private static HashSet<string> GetCards(HashSet<char> group)
        {
            var currentSuits = new HashSet<char>(AllSuits);
            currentSuits.IntersectWith(group);
            if (currentSuits.Count == 0)
                currentSuits = AllSuits;
        
            group.ExceptWith(currentSuits);
            var currentRanks = group;
            if (currentRanks.Count == 0)
                currentRanks = AllRanks;

            var cards = new HashSet<string>();
            foreach (var rank in currentRanks)
            {
                foreach (var suit in currentSuits)
                {
                    cards.Add(rank.ToString() + suit);   
                }
            }

            return cards;
        }
    }
}